### PR TITLE
Chapter 16: Taking Pictures with Intents

### PR DESCRIPTION
**Pull request by OnicaAdrianCatalin**
## New features

 Take photos of crimes.

## Changes in code

 Added an image view and a button.

 Added camera intent.

 Used FileProvider to get images from the camera.

https://user-images.githubusercontent.com/102966412/169258369-d63d0fbe-6ac6-4528-a00f-04d2b4883805.mov

### CODE REVIEW

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 1]
```diff
@@ -167,11 +204,20 @@ class CrimeFragment : Fragment(), FragmentResultListener {
...
-            RESULT_OK -> {
+    private fun onActivityResult(result: ActivityResult, requestCode: Int) {
+        when {
+            result.resultCode != RESULT_OK -> return
+            requestCode == REQUEST_CONTACT -> {
                 onContactSelected(result)
             }
+            requestCode == REQUEST_PHOTO -> {
+                updatePhotoView()
+                requireActivity().revokeUriPermission(
```
Comment (ghost) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt:216]: The book mentions to call this in `onDetach()` too, to cover the possibility of an invalid response

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 2]
```diff
@@ -36,14 +44,20 @@ class CrimeFragment : Fragment(), FragmentResultListener {
...
+    private lateinit var photoButton: ImageButton
+    private lateinit var photoView: ImageView
 
-    private val crimeDetailViewModel: CrimeDetailViewModel by lazy {
+    private val viewModel: CrimeDetailViewModel by lazy {
         ViewModelProvider(this).get(CrimeDetailViewModel::class.java)
     }
 
-    private var resultLauncher =
+    private var resultLauncherSuspect =
```
Comment (ghost) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: can `updatePhotoView` and `resultLauncherPhoto` be `val`s?

**Comment by NegreaVlad** [comment]
>Used FileProvider to get images from the camera

@OnicaAdrianCatalin FileProvider is a provider to a file. Meaning that it provides access to a file. This provider is created by your app.
Can you explain the flow of data? Where is the data stored (in your app or in a space managed by the photo taking activity? Who writes data to the file?
Please explain in a comment the life of the file from the time it is created to the time the photo is displayed in the app.

File: app/build.gradle [hunk 3]
```diff
@@ -40,6 +40,7 @@ dependencies {
     implementation 'androidx.core:core-ktx:1.7.0'
     implementation 'androidx.room:room-runtime:2.4.2'
     implementation 'com.google.android.material:material:1.5.0'
+    implementation 'androidx.window:window:1.1.0-alpha02'
```
Comment (NegreaVlad) [inline app/build.gradle]: Why is this needed?

File: app/src/main/java/com/example/criminalintent/data/local/CrimeDao.kt [hunk 4]
```diff
@@ -18,10 +18,10 @@ interface CrimeDao {
     @Query("SELECT * FROM crime WHERE id =(:id)")
     fun getCrime(id: Int): LiveData<Crime?>
 
-    @Update(onConflict = OnConflictStrategy.REPLACE)
+    @Update(onConflict = OnConflictStrategy.IGNORE)
     fun updateCrime(crime: Crime)
 
-    @Insert(onConflict = OnConflictStrategy.IGNORE)
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/data/local/CrimeDao.kt]: Why is this needed?

File: app/src/main/java/com/example/criminalintent/data/repository/CrimeRepository.kt [hunk 5]
```diff
@@ -31,13 +34,14 @@ class CrimeRepository private constructor(private val crimeDao: CrimeDao) {
         }
     }
 
+    fun getPhotoFile(crime: Crime) = File(filesDir, crime.photoFileName)
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/data/repository/CrimeRepository.kt]: You're only using the photoFileName. Why pass the entire object?

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt (hunk 2 above)
Comment (NegreaVlad) [reply app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: did you run detekt and ktlintCheck?

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 6]
```diff
@@ -146,10 +163,30 @@ class CrimeFragment : Fragment(), FragmentResultListener {
         }
     }
 
+    private fun onCreatePhoto() {
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: A better name IMO `onTakePhotoClicked`

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 7]
```diff
@@ -167,11 +204,20 @@ class CrimeFragment : Fragment(), FragmentResultListener {
         }
     }
 
-    private fun onActivityResult(result: ActivityResult) {
-        when (result.resultCode) {
-            RESULT_OK -> {
+    private fun onActivityResult(result: ActivityResult, requestCode: Int) {
+        when {
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: Interesting when. I find it easier to read if the check if result ok is outside of the when and then the when only checks the `requestCode`

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 8]
```diff
@@ -203,7 +249,12 @@ class CrimeFragment : Fragment(), FragmentResultListener {
                 .setMessage(R.string.dialog_message)
                 .setNegativeButton(
                     R.string.dialog_negative
-                ) { _, _ -> activity?.supportFragmentManager?.popBackStack() }
+                ) { _, _ ->
+                    activity?.supportFragmentManager?.popBackStack()
+                    if (viewModel.crimeLiveData.value?.photoFileName == null) {
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: This logic should not be tied to the dialog. A better place for it is in `onDestroy`.
The code to delete the file should be in viewModel. You can create a method called `releaseResources` or something like that in the view model

**Comment by OnicaAdrianCatalin** [comment]
> > Used FileProvider to get images from the camera
> 
> @OnicaAdrianCatalin FileProvider is a provider to a file. Meaning that it provides access to a file. This provider is created by your app. Can you explain the flow of data? Where is the data stored (in your app or in a space managed by the photo taking activity? Who writes data to the file? Please explain in a comment the life of the file from the time it is created to the time the photo is displayed in the app.

When the file is created it is stored in a file from our app that we configured through provider and meta-data in Android Manifest. After that, the file is written by the camera and is displayed by setting the content of the ImageView with the specific file’s Uri.

**Comment by NegreaVlad** [comment]
Correct! File provider allows other apps to make changes to files under your app's authority

File: app/src/main/java/com/example/criminalintent/data/local/CrimeDao.kt [hunk 9]
```diff
@@ -18,10 +17,10 @@ interface CrimeDao {
     @Query("SELECT * FROM crime WHERE id =(:id)")
     fun getCrime(id: Int): LiveData<Crime?>
 
-    @Update(onConflict = OnConflictStrategy.IGNORE)
+    @Update
     fun updateCrime(crime: Crime)
 
-    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    @Insert
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/data/local/CrimeDao.kt]: Please explain why you removed the `onClinflict` strategy

File: app/src/main/java/com/example/criminalintent/data/model/Crime.kt [hunk 10]
```diff
@@ -11,5 +11,5 @@ data class Crime(
     var date: Date = Date(),
     var isSolved: Boolean = false,
     var suspect: String = "",
-    val photoFileName: String = "IMG_${System.currentTimeMillis()}.jpg"
+    var photoFileName: String = "IMG_${System.currentTimeMillis()}.jpg"
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/data/model/Crime.kt]: Why is this var?

File: app/src/main/java/com/example/criminalintent/data/repository/CrimeRepository.kt [hunk 11]
```diff
@@ -34,10 +34,12 @@ class CrimeRepository private constructor(
         }
     }
 
-    fun getPhotoFile(crime: Crime) = File(filesDir, crime.photoFileName)
+    fun getPhotoFile() = File(filesDir, PHOTO_FILE_NAME)
+
     companion object {
         private var INSTANCE: CrimeRepository? = null
         const val DATABASE_NAME = "crime-database"
+        private const val PHOTO_FILE_NAME = "temporary_file"
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/data/repository/CrimeRepository.kt]: Please specify that this is a temporary file

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 12]
```diff
@@ -108,6 +108,12 @@ class CrimeFragment : Fragment(), FragmentResultListener {
         return when (item.itemId) {
             R.id.add_crime -> {
                 addOrUpdateCrime()
+                viewModel.getPhotoFile().renameTo(
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: Does this work as expected when updating a profile picture?

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 13]
```diff
@@ -205,13 +213,15 @@ class CrimeFragment : Fragment(), FragmentResultListener {
     }
 
     private fun onActivityResult(result: ActivityResult, requestCode: Int) {
-        when {
-            result.resultCode != RESULT_OK -> return
-            requestCode == REQUEST_CONTACT -> {
+        if (result.resultCode != RESULT_OK) {
+            return
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt:208]: Nice :)

**Review by NegreaVlad — COMMENTED** [review]
Please split the changes into smaller commits that address fewer issues (ideally just one issue type i.e releaseResources in view model)
Looks good. Some open questions

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt [hunk 14]
```diff
@@ -108,11 +108,7 @@ class CrimeFragment : Fragment(), FragmentResultListener {
         return when (item.itemId) {
             R.id.add_crime -> {
                 addOrUpdateCrime()
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeFragment.kt]: This method just calls a method from the viewModel. Why not call it directly?

File: app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeDetailViewModel.kt [hunk 15]
```diff
@@ -30,8 +30,25 @@ class CrimeDetailViewModel : ViewModel() {
...
+        crimeLiveData.value?.photoFileName?.let { photoFileName ->
+            getPhotoFile(photoFileName).apply {
+                if (exists() && getPhotoFile().exists()) {
+                    delete()
+                }
+            }
+        }
+        getPhotoFile().renameTo(
+            getPhotoFile(crime.photoFileName)
+        )
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crime_detail/CrimeDetailViewModel.kt:51]: The `addOrUpdate` function now always gets passed the viewModel.crime for argument. Do we still need that parameter to be passed in?

File: app/src/main/java/com/example/criminalintent/presentation/crimeDetail/CrimeDetailViewModel.kt [hunk 16]
```diff
@@ -69,7 +69,7 @@ class CrimeDetailViewModel : ViewModel() {
         crimeRepository.updateCrime(crime)
     }
 
-    fun addOrUpdate(crime: Crime) {
+    fun addOrUpdate() {
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/criminalintent/presentation/crimeDetail/CrimeDetailViewModel.kt]: With the removal of the param, it's not clear what this method is adding or updating. So I would rename to `addOrUpdateCrime` to make things clear, but it's just a suggestion

**Review by NegreaVlad — APPROVED** [review]
LGTM, thanks for the fixes! :)
